freshmaker/pyxis.py

```diff
@@ -0,0 +1,205 @@
+    def status_code(self):
+        return self._status_code
+
+    def __str__(self):
+        return 'Errors ({}):\n{}'.format(
+            len(self._raw['errors']),
+            '\n'.join(('    {}'.format(err['msg'])
+                      for err in self._raw['errors']))
+        )
+
+
+class Pyxis(object):
+    """ Interface for querying Pyxis"""
+
+    def __init__(self, server_url):
+        self._server_url = server_url.rstrip('/')
+        # add api version to root url
+        self._api_root = f"{self._server_url}/v1"
+        # list of default fields to include in request to pyxis
+        self._default_include_fields = ['page_size', 'total']
+
+    def _make_request(self, entity, params):
+        """ Send a request to Pyxis
+
+        :param str entity: entity part to construct a full URL for request.
+        :param dict params: Pyxis query parameters.
+        :return: Json response from Pyxis
+        :rtype: dict
+        :raises PyxisRequestError: If Pyxis returns error as a response
+        """
+        entity_url = '{}/{}'.format(self._api_root, entity)
+        default_params = {"page_size": "100"}
+        default_params.update(params)
+
+        response = requests.get(entity_url, params=default_params,
+                                auth=HTTPKerberosAuth(mutual_authentication=OPTIONAL))
+
+        if response.status_code == HTTPStatus.OK:
+            return response.json()
+
+        # Warn early, in case there is an error in the error handling code below
+        log.warning("Request to %s gave %r", response.request.url, response)
+
+        raise PyxisRequestError(response.status_code, response.json())
+
+    def _pagination(self, entity, params):
+        """
+        Process all pages in Pyxis
+
+        :param str entity: what data/entity to request from Pyxis
+        :param dict params: parameters to add to GET request
+        :return: list of all 'data' fields from responses from Pyxis
+        :rtype: list
+        """
+        ret = []
+        response_data = self._make_request(entity, params=params)
+        ret.extend(response_data.get('data', []))
+        page = 1
+        last_page = ceil(response_data['total'] / response_data['page_size'])
+        while page < last_page:
+            params.update({"page": page})
+            response_data = self._make_request(entity, params=params)
+            ret.extend(response_data.get('data', []))
+            page += 1
+        return ret
+
+    def get_operator_indices(self, params={}):
+        """
+        Get all index images for organization(s)(configurable) from Pyxis
+
+        :param dict params: parameters to add to GET request
+        """
+        organization = conf.pyxis_index_image_organization
+        params.update({"filter": "organization=="+organization})
+
+        entity = "operators/indices"
+        indices = self._pagination(entity, params)
+        return indices
+
+    def _get_bundles_per_index_image(self, index_images, params={}):
```

Comment:
In general, you shouldn't use non-primitive types as defaults for function/method keyword argument defaults. This example shows you why:
```python
>>> def example(params={}):
...     print(params)
...     params['message'] = "Will this persist in the next function call?"
... 
>>> example()
{}
>>> example()
{'message': 'Will this persist in the next function call?'}
```